_posts/2017-12-08-totw-123.md

```diff
@@ -0,0 +1,196 @@
+    difficult if complicated expressions are involved.
+
+### As `absl::optional<Bar>`
+
+This is a good middle ground between the simplicity of bare objects and the
+flexibility of `std::unique_ptr`. The object is stored inside `Foo` but, unlike
+bare objects, `absl::optional` can be empty. It can be populated at any time by
+assignment (`opt_ = ...`) or by constructing the object in place
+(`opt_.emplace(...)`).
+
+Because the object is stored inline, the usual caveats about allocating large
+objects on the stack apply, just like for a bare object. Also be aware that an
+empty `absl::optional` uses as much memory as a populated one.
+
+Compared to a bare object, `absl::optional` has a few downsides:
+
+*   It's less obvious for the reader where object construction and destruction
+    occur.
+*   There is a risk of accessing an object which does not exist.
+
+### As `std::unique_ptr<Bar>`
+
+This is the most flexible way. The object is stored outside of `Foo`. Just like
+`absl::optional`, a `std::unique_ptr` can be empty. However, unlike
+`absl::optional`, it is possible to transfer ownership of the object to
+something else (through a move operation), to take ownership of the object from
+something else (at construction or through assignment), or to assume ownership
+of a raw pointer (at construction or through `ptr_ = absl::WrapUnique(...)`, see
+[TotW 126](/tips/126).
+
+When `std::unique_ptr` is null, it doesn't have the object allocated, and
+consumes only the size of a pointer[^deleter].
+
+Wrapping an object in a `std::unique_ptr` is necessary if the object may need to
+outlive the scope of the `std::unique_ptr` (ownership transfer). 
+
+This flexibility comes with some costs:
+
+*   Increased cognitive load on the reader:
+    *   It's less obvious what's stored inside (`Bar`, or something derived from
+        `Bar`). However, it may also decrease the cognitive load, as the reader
+        can focus only on the base interface held by the pointer.
+    *   It's even less obvious than with `absl::optional` where object
+        construction and destruction occur, because ownership of the object can
+        be transferred.
+*   As with `absl::optional`, there is a risk of accessing an object which does
+    not exist - the famous null pointer dereference.
+*   The pointer introduces an additional level of indirection, which requires a
+    heap allocation, and is [not
+    friendly](https://en.wikipedia.org/wiki/Locality_of_reference) to CPU
+    caches; Whether this matters or not depends a lot on particular use cases.
+*   `std::unique_ptr<Bar>` is not copyable even if `Bar` is. This also prevents
+    `Foo` from being copyable. 
+
+### Conclusion
+
+As always, strive to avoid unnecessary complexity, and use the simplest thing
+that works. Prefer bare object, if it works for your case. Otherwise, try
+`absl::optional`. As a last resort, use `std::unique_ptr`.
+
+<table>
+  <tr>
+    <td>`Bar`</td>
```

Comment:
You need an opening blank <td></td> here to hold the first empty cell and shift the headers one cell each.